### PR TITLE
Fix rendering of models with uncentered icons

### PR DESCRIPTION
**Pull request by MaksKBV**
### Related Issues

OMEdit incorrectly renders a fliped model with uncentered icon #11250

### Purpose

OMEdit incorrectly renders a fliped model with uncentered icon

### Approach

Change getTransformationMatrixDiagram function to correctly form the QTransform

### CODE REVIEW

File: OMEdit/OMEditLIB/Element/Element.cpp [hunk 1]
```diff
@@ -554,8 +554,10 @@ Element::Element(ModelInstance::Component *pModelComponent, bool inherited, Grap
     ModelInstance::CoordinateSystem coordinateSystem = getCoOrdinateSystemNew();
     qreal initialScale = coordinateSystem.getInitialScale();
     QVector<QPointF> extent;
-    extent.append(QPointF(initialScale * boundingRect().left(), initialScale * boundingRect().top()));
-    extent.append(QPointF(initialScale * boundingRect().right(), initialScale * boundingRect().bottom()));
+    qreal xExtent = initialScale * boundingRect().width() / 2;
+    qreal yExtent = initialScale * boundingRect().height() / 2;
+    extent.append(QPointF(-xExtent, -yExtent));
+    extent.append(QPointF(xExtent, yExtent));
```
Comment (adeas31) [inline OMEdit/OMEditLIB/Element/Element.cpp:560]: This seems to be doing the same thing. We might need to consider the case when either width or height is zero.

File: OMEdit/OMEditLIB/Element/Element.cpp [hunk 2]
```diff
@@ -3443,16 +3447,26 @@ void Element::resizeElement(QPointF newPosition)
...
+  qreal x = pivot.x() + (mTransformation.getOrigin().x() - pivot.x()) * kx;
+  qreal y = pivot.y() + (mTransformation.getOrigin().y() - pivot.y()) * ky;
+  ExtentAnnotation currentExtent = mTransformation.getExtent();
+  QPointF currentExtent1 = currentExtent.at(0);
+  QPointF currentExtent2 = currentExtent.at(1);
+  extent1.setX(currentExtent1.x() * kx);
+  extent1.setY(currentExtent1.y() * ky);
+  extent2.setX(currentExtent2.x() * kx);
+  extent2.setY(currentExtent2.y() * ky);
+  mTransformation.setOrigin(QPointF(x, y));
```
Comment (adeas31) [inline OMEdit/OMEditLIB/Element/Element.cpp:3469]: Something is off with this code. Try resizing the components. Resize them in a way that they flip horizontally and vertically.

**Review by adeas31 — CHANGES_REQUESTED** [review]
The code works in most of the cases. However, it doesn't work well when resizing the components and specially when resizing the rotated components. Also the flipping and rotation is done based on the origin and not the component center.

**Comment by MaksKBV** [comment]
I did the flipping and rotation based on the origin. I will check your other comments.

File: OMEdit/OMEditLIB/Element/Element.cpp (hunk 1 above)
Comment (MaksKBV) [reply OMEdit/OMEditLIB/Element/Element.cpp:560]: I try this code with models with {{0,0},{0,0}} icon and diagram extents and it works fine

**Comment by MaksKBV** [comment]
I try rotate, flip, resize combinations on new version and it works fine for me. When you resize origin moves proportional relative to fixed point. If it was in the center it will be in the center after resize.

**Comment by MaksKBV** [comment]
Oh, ok i see problem when resize component through origin

**Comment by MaksKBV** [comment]
Fix resize. Now code calculete new coordinates from start origin and extent. So it works fine even when pass point with zero square.

**Review by adeas31 — CHANGES_REQUESTED** [review]
Thanks for the fixes. The rotation and flipping works fine now. The resize still has some issues. For example,

Take the following code,

```
model M
  Modelica.Electrical.Analog.Basic.Ground ground annotation (
    Placement(transformation(origin = {5, 5}, extent = {{-15, -15}, {15, 15}}, rotation = -90)));
equation 

annotation (
    uses(Modelica(version = "4.0.0")));
end M;
```

Resize ground using the top right red corner. Move it down so that the ground is flipped vertically. Instead the ground is flipped horizontally and the resulting Modelica code is wrong. You have to take the rotation in consideration when resizing.

**Comment by adeas31** [comment]
Also what is done for `Transformation::getTransformationMatrixDiagram()`. Do the same in `Transformation::getTransformationMatrixIcon()`.

**Comment by MaksKBV** [comment]
Added angle inversion when flip with the mouse. Did the same for icons.

**Review by adeas31 — APPROVED** [review]

**Pull request merged**